### ISSUE TITLE
Methodology for Validation

### ISSUE THREAD

**Issue by samhatchett** [OpenWaterAnalytics/EPANET#169]
In addition to regression and unit testing, I believe the project could benefit from integrating a statistical validation procedure into the mix.

There are some references in other fields - like from [Hassan] (https://www.lm.doe.gov/CNTA/CNT2004.pdf) and [Bennett](https://www.sciencedirect.com/science/article/pii/S1364815212002435) 

A basic strategy would be to have a set of `inp` files that describe experimental setups, and provide real measurement data from field evaluations. Then use peer-reviewed algorithms like the wonderfully-named "[GLUE](https://www.sciencedirect.com/science/article/pii/S0022169401004218)" (Generalized Likelihood Uncertainty Estimator) or the perhaps easier to parse Stochastic Validation Approach in [Luis and McLaughlin](https://www.sciencedirect.com/science/article/pii/0309170892900292) to test.

In all, this would go some way to helping get around the questions of reproducibility - namely, that regression testing also constrains you to reproduce errors present in past versions of code, rather than working toward better statistical validity which is the real point.

A special note - I have to credit Rakesh Bahadur at Leidos for directing me to this research and encouraging us to pursue this validation approach. My hat is off to him for seeing the bigger picture.

**Comment by michaeltryby** [comment]
Thinking about it more, we should be able to develop an nrtest plugin to do validations as well as regression tests. But perhaps there are better tools our there for doing this.

**Comment by LRossman** [comment]
Please don't conflate the notions of model verification and model validation. They are two different concepts. See [https://en.wikipedia.org/wiki/Verification_and_validation_of_computer_simulation_models](url).

**Comment by samhatchett** [comment]
Thanks @LRossman - I suppose this deserves a fuller conversation, but I totally agree. Verification can be done through unit testing, but validation must be done statistically, right? Do you have any historical perspective on either question as it relates to EPANET?

**Comment by michaeltryby** [comment]
Yes I agree, regression testing is a method for model verification and is different than model validation. We use nrtest for model validation using a custom plugin for reading EPANET binary files. Nrtest provides a framework for running models and managing inputs and outputs. The plugin provides the comparison routine. 

We could also write an nrtest plugin that would focus more on model validation. Such a plugin could read in measured data sets and compare them against model results using a statistical approach. That is to say we could reuse some of the same software to perform model verification and model validation to the extent that the mechanics of the use cases are similar even though the use cases are different and should no be conflated with one another.

**Comment by LRossman** [comment]
With regards to EPANET's hydraulic solver, verification can be built in to the code itself in the form of having it report the error in meeting the conservation of energy and flow equations it is solving (although I suppose a regression test of new versions is still needed to confirm that these errors are being computed correctly). With water quality, an internal verification check would be the reporting of mass balance errors.

Model validation is a tougher nut to crack since it depends on so many uncontrollable factors. Plus there is a philosophical belief that models of open systems can never be fully validated (see the classic paper by Orestes et al. [https://pdfs.semanticscholar.org/c75d/9457985303b84d3726440151a46d7135a428.pdf](url) - don't be confused by their switching the definitions of verification and validation). Instead of trying to show that a model is valid, statistically or otherwise, it might be more productive to provide empirical evidence that it provides useful information for its intended use.

**Comment by dickinsonre** [comment]
Hello Lew, Is that the right link for the Orestes et al paper?

**Comment by LRossman** [comment]
Re: the Orestes et al. paper - if left clicking on the lnk doesn't work, try right clicking it and select "Search Google for "....'".

**Comment by dickinsonre** [comment]
Thanks @LRossman that your suggestion worked for me and I now have the paper.
